server/integration/infinispan/src/main/java/org/infinispan/server/infinispan/task/DistributedServerTask.java

```diff
@@ -0,0 +1,56 @@
+package org.infinispan.server.infinispan.task;
+
+import org.infinispan.Cache;
+import org.infinispan.commons.marshall.Marshaller;
+import org.infinispan.distexec.DistributedCallable;
+import org.infinispan.factories.GlobalComponentRegistry;
+import org.infinispan.tasks.TaskContext;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Author: Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ * Date: 1/28/16
+ * Time: 1:49 PM
+ */
+public class DistributedServerTask<T> implements Serializable, DistributedCallable<Object, Object, T> {
+   private final Optional<Map<String, ?>> parameters;
+   private final String taskName;
+
+   private transient GlobalComponentRegistry componentRegistry;
+   private transient Cache<Object, Object> cache;
+
+   public DistributedServerTask(String taskName, Optional<Map<String, ?>> parameters) {
+      this.taskName = taskName;
+      this.parameters = parameters;
+   }
+
+   @Override
+   public void setEnvironment(Cache<Object, Object> cache, Set<Object> inputKeys) {
+      this.cache = cache;
+      // todo script-related code wraps a lot of things in security util calls. is this necessary here?
+      componentRegistry = cache.getCacheManager().getGlobalComponentRegistry();
+   }
+
+   @Override
+   public T call() throws Exception {
+      ServerTaskRegistry taskRegistry = componentRegistry.getComponent(ServerTaskRegistry.class);
```

Comment:
I think the ServerTaskRegistry should not be obtained here, but it should be injected by the runner during construction.
Also, a cache is optional paramater in the TaskContext and you should not rely on its presence here. Because of a current limitation in the DefaultExecutorService, it is not possible to have a distexec without a distributed cache, but @wburns is adding a new implementation which changes this.